lib/base/bulker.hpp

```diff
@@ -0,0 +1,116 @@
+/* Icinga 2 | (c) 2022 Icinga GmbH | GPLv2+ */
+
+#ifndef BULKER_H
+#define BULKER_H
+
+#include <boost/config/detail/suffix.hpp>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <utility>
+#include <vector>
+
+namespace icinga
+{
+
+/**
+ * A queue which outputs the input as bulks of a defined size
+ * or after a defined time, whichever is reached first
+ *
+ * @ingroup base
+ */
+template<class T>
+class Bulker
+{
+private:
+	typedef std::chrono::steady_clock Clock;
+
+public:
+	typedef std::vector<T> Container;
+	typedef typename Container::size_type SizeType;
+	typedef typename Clock::duration Duration;
+
+	Bulker(SizeType bulkSize, Duration threshold)
+		: m_BulkSize(bulkSize), m_Threshold(threshold), m_NextConsumption(NullTimePoint()) { }
+
+	void ProduceOne(T needle);
+	Container ConsumeMany();
+
+private:
+	typedef std::chrono::time_point<Clock> TimePoint;
+
+	static inline
+	TimePoint NullTimePoint()
+	{
+		return TimePoint::min();
+	}
+
+	inline void UpdateNextConsumption()
+	{
+		m_NextConsumption = Clock::now() + m_Threshold;
+	}
+
+	const SizeType m_BulkSize;
+	const Duration m_Threshold;
+
+	std::mutex m_Mutex;
+	std::condition_variable m_CV;
+	std::queue<Container> m_Bulks;
+	TimePoint m_NextConsumption;
+};
+
+template<class T>
+void Bulker<T>::ProduceOne(T needle)
+{
+	std::unique_lock<std::mutex> lock (m_Mutex);
+
+	if (m_Bulks.empty() || m_Bulks.back().size() == m_BulkSize) {
+		m_Bulks.emplace();
+	}
+
+	m_Bulks.back().emplace_back(std::move(needle));
+
+	if (m_Bulks.size() == 1u && m_Bulks.back().size() == m_BulkSize) {
+		m_CV.notify_one();
+	}
+}
+
+template<class T>
+typename Bulker<T>::Container Bulker<T>::ConsumeMany()
+{
+	std::unique_lock<std::mutex> lock (m_Mutex);
+
+	if (BOOST_UNLIKELY(m_NextConsumption == NullTimePoint())) {
+		UpdateNextConsumption();
+	}
+
+	for (;;) {
+		if (!m_Bulks.empty() && m_Bulks.front().size() == m_BulkSize) {
+			auto haystack (std::move(m_Bulks.front()));
+
+			m_Bulks.pop();
+			UpdateNextConsumption();
+			return std::move(haystack);
+		}
+
+		auto deadline (m_NextConsumption);
+
+		if (m_CV.wait_until(lock, deadline) == std::cv_status::timeout) {
+			if (m_Bulks.empty()) {
+				UpdateNextConsumption();
+				return Container();
+			}
+
+			auto haystack (std::move(m_Bulks.front()));
+
+			m_Bulks.pop();
+			UpdateNextConsumption();
+			return std::move(haystack);
```

Comment:
"should not be"? May it be or not? If definitely not, I lazy piece of meat would of course save the additional 11 characters to type. :)

Reply:
To be honest:

* #7593
* https://www.youtube.com/watch?v=Aul35gjdFWM (0-10s)

:)

Reply:
The start of your comment sounds like I'll just let it as-is because... well, see your 1st sentence. And the comment end.. should moving vs. directly construction matter?

Reply:
It's a `std::vector`, so for a typical implementation on a 64 bit platform, this will copy an additional 24 bytes. Nothing that will kill us, but still, if you ask the compiler (`-Wpessimizing-move`), it explicitly tells you to not move here because it prevents an optimization.